storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java

```diff
@@ -393,14 +400,15 @@ public void refreshConnections() {
         Set<NodeInfo> newConnections = Sets.difference(neededConnections, currentConnections);
         Set<NodeInfo> removeConnections = Sets.difference(currentConnections, neededConnections);
 
+        Map<String, String> nodeHost = assignment != null ? assignment.get_node_host() : null;
         // Add new connections atomically
         cachedNodeToPortSocket.getAndUpdate(prev -> {
             Map<NodeInfo, IConnection> next = new HashMap<>(prev);
             for (NodeInfo nodeInfo : newConnections) {
                 next.put(nodeInfo,
                          mqContext.connect(
                              topologyId,
-                             assignment.get_node_host().get(nodeInfo.get_node()),    // Host
+                             nodeHost.get(nodeInfo.get_node()),    // Host
```

Comment:
It could throw NPE when assignment is null since nodeHost will be null. If there's some assumption this line will not be evaluated when assignment is null, may be better to leave a comment describing it.

Reply:
Added a comment. NodeHost can't be null here, since we're iterating over newConnections. newConnections is only non-empty if assignment is not null.

I looked at rewriting this to make that clearer, but didn't like how it turned out.

Reply:
That's OK once we make clear it doesn't throw NPE. Thanks for explaining here.